### ISSUE TITLE
Collection notifications shouldn't fire until changes are actually made

### ISSUE THREAD

**Issue by alazier** [realm/realm-js#950]
Right now they fire with an empty notification when registered. This is confusing and annoying.

**Comment by fealebenpae** [comment]
This is Object Store behavior.

**Comment by alazier** [comment]
Yeah - but can/should we suppress this? Would this change in general be an improvement? Is there any reason we want an initial empty notification?

**Comment by fealebenpae** [comment]
I think we should change this behavior at the Object Store level - we've also found the initial empty notification vexing in dotnet.

That said, removing the initial notification would be a breaking change.

**Comment by blagoev** [comment]
👍 for fixing this at the source. Trying to hide it ourselves will add more to the confusion.

**Comment by bdash** [comment]
The initial notification provides the current state of the collection. Delivering the initial results asynchronously allows the query to be evaluated on a background thread, rather than blocking the main thread.

**Comment by alazier** [comment]
@bdash I think this is fine if we don't return the Results object until the collection is ready to be accessed. But with our current apis we return the Results object synchronously and then impose the constraint that users don't interact with it until the callback is called the first time. This is kind of confusing and error prone. I guess we could fix this in each binding's usage of Results but we should probably reach consensus on if this is a change we actually want to make.

**Comment by bdash** [comment]
There's nothing preventing a user from interacting with it prior to the callback being called. It's just that if they do access it prior to the callback being called to deliver the initial results, we'll evaluate the query synchronously to provide the results.

> This is kind of confusing and error prone. I guess we could fix this in each binding's usage of Results but we should probably reach consensus on if this is a change we actually want to make.

I guess I can see how it may be confusing if you're unfamiliar with the design, but I'm not sure what you think is error-prone about it, nor what bindings would need to fix.

**Comment by alazier** [comment]
It's error prone as if users don't understand the design they may inadvertently block the calling thread. The fix would be to return the Results object asynchronously when they are first available rather than returning an unevaluated Results immediately.

**Comment by mdekrey** [comment]
From the documentation on collection notifications:

> Collection notifications are delivered asynchronously: first with the initial results, and then after any write transaction which modifies any of the objects in the collection, deletes objects from the collection, or adds new objects to the collection.

This is useful in a stream-of-events to build and maintain a list changing only the values that actually changed - starting with an empty list. For people who use it as documented (and as @bdash describes) for asynchronous results, this could be a significant breaking change.
